libs/wingsdk/src/target-sim/function.inflight.ts

```diff
@@ -78,12 +96,94 @@ export class Function implements IFunctionClient, ISimulatorResourceInstance {
               sourceType: FUNCTION_TYPE,
               timestamp: new Date().toISOString(),
             });
-          });
+          };
         }
 
-        const index = vm.runFile(this.filename);
-        return index.handler(payload);
+        return runSandbox(this.bundle, payload, {
+          timeout: this.timeout,
+          context: {
+            process: sandboxProcess,
+            $simulator: this.context,
+            console: sandboxConsole,
+          },
+        });
       },
     });
   }
 }
+
+interface RunSandboxOptions {
+  readonly context: { [key: string]: any };
+  readonly timeout: number;
+}
+
+/**
+ * Runs user code in a sandboxed environment. The code is expected to export a `handler` async
+ * function which take a payload and returns a result via a promise.
+ *
+ * @param filepath A path to a bundled JavaScript file (no "require")
+ * @param payload The payload JSON object to pass to the handler
+ * @param opts Sandbox options
+ * @returns A promise
+ */
+async function runSandbox(
+  filepath: string,
+  payload: any,
+  opts: RunSandboxOptions
+): Promise<any> {
+  const ctx: any = {};
+
+  // create a copy of all the globals from our current context.
+  for (const k of Object.getOwnPropertyNames(global)) {
+    try {
+      ctx[k] = (global as any)[k];
+    } catch {
+      // ignore unresolvable globals (see https://github.com/winglang/wing/pull/1923)
+    }
+  }
+
+  // append the user's context
+  for (const k of Object.keys(opts.context)) {
+    ctx[k] = opts.context[k];
+  }
+
+  const code = await readFile(filepath, "utf-8");
+
+  return new Promise(($resolve, $reject) => {
+    const wrapper = [
+      "const exports = {};",
+      code,
+      `exports.handler(${JSON.stringify(
+        payload
+      )}).then($resolve).catch($reject);`,
+    ].join("\n");
+
+    const context = vm.createContext({
+      ...ctx,
+      $resolve,
+      $reject,
+      require, // to support requiring node.js sdk modules (others will be bundled)
+    });
+
+    // emit an explicit error when trying to access `__dirname` and `__filename` because we cannot
+    // resolve these when bundling (this is true both for simulator and the cloud since we are
+    // bundling there as well).
+    const forbidGlobal = (name: string) => {
+      Object.defineProperty(context, name, {
+        get: () => {
+          throw new Error(
+            `${name} cannot be used within bundled cloud functions`
+          );
+        },
+      });
+    };
+
+    forbidGlobal("__dirname");
+    forbidGlobal("__filename");
```

Comment:
> Unfortunately, modules that use __dirname and/or __filename cannot be bundled and therefore are not supported. We emit an explicit error when these are used.

We're using __dirname / __filename a few places in the SDK, I'm curious to understand how those code paths are still supported (I'd have expected those usages might have to be removed)? Are those usages okay because they're only in preflight code?

Reply:
Yes, in preflight there's no problem because we are not bundling.